pod-overhead: add documentation (#15970)
kubernetes/website: content/en/docs/concepts/containers/pod-overhead.md

> ---

Reviewer: I would expect to see this filed inside https://kubernetes.io/docs/concepts/workloads/pods/

Reviewer 2: Or maybe https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/ makes more sense?

Author: agreed.  changed accordingly. PTAL

reviewers:
- tallclair
- dchen1107
- egernst
title: Pod Overhead
content_template: templates/concept
weight: 20
---

{{% capture overview %}}

> {{< feature-state for_k8s_version="v1.16" state="alpha" >}}

Reviewer 2: That's correct. I think it's good to call that out and link to the feature gates doc from here.

> This page describes pod Overhead, as defined in RuntimeClass and pod.Spec, and mechanisms for applying this feature.

Reviewer 2: nit: inconcistent capitalization. Stick with either `Pod Overhead` or `pod overhead`

Reviewer: Use italics for the first mention of “Pod Overhead”. Here's a full example that frames _Pod Overhead_ as a general concept with an implementation in Kubernetes.
```markdown
When you run a Pod on a Node, the Pod itself takes an amount of system resources. These
resource are additional to the resources needed to run the container(s) inside the Pod.
_Pod Overhead_ is a way of accounting for the extra resources consumed by the Pod itself.

Kubernetes can track the overhead for Pods, and account for these resources. In Kubernetes,
a Pod's overhead is set at admission time, based on the overhead settings for that
Pod's [RuntimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/).
```

{{% /capture %}}

{{% capture body %}}

## Pod Overhead

> Pod Overhead is a feature for applying an overhead to a pods definition at admission time depending on the RuntimeClass being utilized.

Reviewer: This line feels like a great fit for the overview section.

Also, how about adding a hyperlink to “admission”:
```suggestion
Pod Overhead is a feature for applying an overhead to a pods definition at [admission](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks) time, depending on the RuntimeClass being utilized.
```

If the link doesn't feel right, maybe I can find or tweak another page that explains _admission_ as a Kubernetes concept.

> Pod Overhead is a feature for applying an overhead to a pods definition at admission time depending on the RuntimeClass being utilized.

Reviewer 2: I'd start by explaining what overhead is, and why it's important. Maybe:
```suggestion
Pod Overhead is a feature for accounting for the resources consumed by the pod infrastructure on top of the container requests & limits. The pod's overhead is set at admission time according to the overhead associated with the pod's [RuntimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/).
```

### Set Up

> Ensure the PodOverhead feature gate is enabled (it is by default). See [Feature

Reviewer 2: Alpha features are off by default.
```suggestion
Ensure the PodOverhead feature gate is enabled (it is disabled by default). See [Feature
```

Gates](/docs/reference/command-line-tools-reference/feature-gates/) for an explanation of enabling
feature gates. The `PodOverhead` feature gate must be enabled on scheduler, apiservers _and_ kubelets.

1. Configure basic RuntimeClass
2. Add a pod Overhead to the defined RuntimeClass

> #### 1. Configure basic RuntimeClass

Reviewer: I'd consider stopping the Concept page at this point, and hyperlinking to a new, separate Task page about using PodOverhead.

(it's not a problem for Concept pages to be short; my gut feeling is that that's actually better)

> #### 2. Add pod Overhead to RuntimeClass definition

Reviewer 2: Can you also add a section to the RuntimeClass page? 1-2 sentences plus a link to this document should be sufficient.

{{< note >}}

> It is recommended that RuntimeClass write operations (create/update/patch/delete) be

Reviewer: How about this wording, or something like it:
```markdown
Users who can write to RuntimeClass resources are able to have cluster-wide impact on
workload performance. You can limit access to this ability using Kubernetes access controls.
See …
```
(aim here is to avoid writing in the passive voice&mdash;“It is recommended”).

restricted to the cluster administrator. This is typically the default. See [Authorization
Overview](/docs/reference/access-authn-authz/authorization/) for more details.
{{< /note >}}

> 

Reviewer: ```suggestion
{{% /capture %}}
```
(for tidiness)

> 

Reviewer 2: Can we add a link to the KEP to learn more at the end of the page? Is that something that is done? (@sftim)

Reviewer: Linking to the KEP is relevant, and a `whatsnext` section is a great place to put relevant links.
